File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/IDCardToolkitObjCApp/IDCardToolkitObjCApp/ViewController/PublicDataEFViewController.m
//
//  PublicDataEFViewController.m
//  IDCardToolkitObjCApp
//
//  Created by Federal Authority For Identity and Citizenship on 20/11/19.
//  Copyright © 2019 Federal Authority For Identity and Citizenship. All rights reserved.
//

#import "PublicDataEFViewController.h"

@interface PublicDataEFViewController () {
    NSMutableArray *listArr;
    int selectedEFType;
}
@property(nonatomic,strong)Utils * utils;
@end

@implementation PublicDataEFViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
       self.title=@"Public Data EF";
       self.navigationItem.hidesBackButton = YES;
       self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[[UIImage imageNamed:@"backbutton"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal]
                                                                                style:UIBarButtonItemStylePlain
                                                                               target:self
                                                                               action:@selector(goBack)];
       self.readPublicDataEFButton.layer.cornerRadius=22.0;
       [self.readPublicDataEFButton.layer setShadowOffset:CGSizeMake(5, 5)];
       [self.readPublicDataEFButton.layer setShadowColor:[[UIColor lightGrayColor] CGColor]];
       [self.readPublicDataEFButton.layer setShadowOpacity:0.5];
      // self.readPublicDataEFButton.layer.borderWidth=1.0;
       
       AppDelegate *aDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
       self.model = aDelegate.model;
       self.utils=aDelegate.utils;
   
    listArr=[[NSMutableArray alloc]initWithObjects:@"IDN_CN",@"ROOT_CERTIFICATE",@"NON_MODIFIABLE_DATA",@"MODIFIABLE_DATA",@"PHOTOGRAPHY",@"SIGNATURE_IMAGE",@"HOME_ADDRESS",@"WORK_ADDRESS", nil];
    // Do any additional setup after loading the view.
}

- (IBAction)readPublicDataEFButtonAction:(id)sender {
    
    PUBLIC_DATA_EF_TYPE publicDataEFType = selectedEFType+1;
   // NSLog(@"publicDataEFType %i",publicDataEFType);
     [self.utils ShowProgressBar:LOADINGDATA andView:self.view];
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void) {
            @try {
                CardReader *cardreader = [self.model.cardReaderSharedArray objectAtIndex:0];
                
                [cardreader readPublicDataEF:publicDataEFType validateSignature:true];
                
              NSString *getParsedData =   [cardreader parseEFData:[cardreader getReadPublicDataEFByte] efDatalength:[cardreader getReadPublicDataEFLength]];
            dispatch_async(dispatch_get_main_queue(), ^{
                if ([cardreader getReadPublicDataEFLength]>0) {
                    
                    NSMutableString * getEFDataString = [[NSMutableString alloc] init];
                    
                    for (int i=0; i<[cardreader getReadPublicDataEFLength]; i++) {
                        [getEFDataString appendString:[NSString stringWithFormat:@"%02x",[cardreader getReadPublicDataEFByte][i]]];
                    }
                 //   NSLog(@"parseddata %@",parseddata);

                    [AlertView showAlertTitle:@"" withMessage:[NSString stringWithFormat:@"Parsed EF data \n\n %@",getParsedData] onView:self];
                }
              });// update UI main queue
            }// try
            @catch (NSException *exception) {
                dispatch_async(dispatch_get_main_queue(), ^{
                NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
                    [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
                });// update UI main queue
            }// catch
            @finally {
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self.utils DismissProgressBar];
                });// update UI main queue
            }// finally
        });// background queue
}

#pragma mark PickerView Delagates methods

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}
- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return listArr.count;
}
- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    selectedEFType = (int)row;
    return listArr[row];
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    selectedEFType = (int)row;
}

- (void)goBack {
    [self.navigationController popViewControllerAnimated:YES];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/


@end

File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/ICASwiftApp/ICASwiftApp/PublicDataEFViewController.swift
// FIXME: PUBLIC_DATA_EF_TYPE enum does not bridge to Swift in Xcode 16+.
#if false
//
//  PublicDataEFViewController.swift
//  ICASwiftApp
//
//  Created by Digital Trust Tech on 06/05/22.
//

import UIKit

class PublicDataEFViewController: UIViewController,UIPickerViewDelegate,UIPickerViewDataSource  {
    
    @IBOutlet weak var pickerViewEFData: UIPickerView!
    @IBOutlet weak var readPublicDataEFButton: UIButton!
    @IBOutlet weak var view_Roundedcorners: UIView!
    
    var listArr: [Any] = []
    var selectedEFType = NSInteger()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        
        self.view_Roundedcorners.cornered_View([.layerMaxXMaxYCorner, .layerMinXMaxYCorner], radius: 20, borderColor: .red, borderWidth: 0)
        
        readPublicDataEFButton.layer.cornerRadius = 22.0
        readPublicDataEFButton.layer.shadowOffset = CGSize(width: 5.0, height: 5.0)
        readPublicDataEFButton.layer.shadowColor = UIColor.lightGray.cgColor
        readPublicDataEFButton.layer.shadowOpacity = 0.5
        
        listArr = ["IDN_CN", "ROOT_CERTIFICATE", "NON_MODIFIABLE_DATA", "MODIFIABLE_DATA", "PHOTOGRAPHY", "SIGNATURE_IMAGE", "HOME_ADDRESS", "WORK_ADDRESS"]
    }
    
    @IBAction func btn_BackAction(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }
    
    @IBAction func readPublicDataEFButtonAction(_ sender: Any) {
        
        if toolkit == nil  {
            self.view.makeToast(Constants.toolkitNotInitilized)
            return
        }//if
        else if cardreader == nil {
            self.view.makeToast(Constants.cardreaderNotConnect)
            return
        }//else if
        
        let publicDataEFType = (PUBLIC_DATA_EF_TYPE.init(UInt32(selectedEFType)).rawValue)+1
        
        utils.showProgressBar(Constants.LoadingData, andView: self.view)
        DispatchQueue.global(qos: .background).async {
            do {
                try cardreader?.readPublicDataEF(PUBLIC_DATA_EF_TYPE(rawValue: publicDataEFType), validateSignature: 1)
                let length = cardreader?.getreadPublicDataEFLength()
                let byteData = cardreader?.getreadPublicDataEFByte()
                
                let getParsedData =  try cardreader?.parseEFData(efData: byteData, efDatalength:UInt32(length!))
                
                DispatchQueue.main.async {
                    
                    //                    var getEFDataString = ""
                    //                    for i in 0..<length {
                    //                        getEFDataString += String(format: "%02x",byteData?[i] as! UInt8)
                    //                    }
                    AlertView.showAlertTitle("", withMessage:"Parsed EF data \n\n  \(getParsedData!)",onView: self)
                    
                    utils.dismissProgressBar()
                }// update UI main queue
            }//do
            catch let error as NSError {
                DispatchQueue.main.async {
                    let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                    AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                    utils.dismissProgressBar()
                }// update UI main queue
            }//catch
        }//background queue
    }
    
    
    // MARK:PickerView Delagates methods
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return listArr.count
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        selectedEFType = row
        return listArr[row] as? String
    }
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        selectedEFType = row
    }
    @objc func goBack() {
        navigationController?.popViewController(animated: true)
    }
}
#endif
